components/script/dom/document.rs

```diff
@@ -1448,8 +1448,73 @@ impl Document {
         event.fire(target, can_gc);
     }
 
+    /// <https://www.w3.org/TR/clipboard-apis/#clipboard-actions>
+    pub fn handle_clipboard_action(&self, action: ClipboardEventType, can_gc: CanGc) -> bool {
+        // The script_triggered flag is set if the action runs because of a script, e.g. document.execCommand()
+        let script_triggered = false;
+
+        // The script_may_access_clipboard flag is set
+        // if action is paste and the script thread is allowed to read from clipboard or
+        // if action is copy or cut and the script thread is allowed to modify the clipboard
+        let script_may_access_clipboard = false;
+
+        // Step 1 If the script-triggered flag is set and the script-may-access-clipboard flag is unset
+        if script_triggered && !script_may_access_clipboard {
+            return false;
+        }
+
+        // Step 2 Fire a clipboard event
+        let event = ClipboardEvent::new(
+            &self.window,
+            None,
+            DOMString::from(action.as_str()),
+            EventBubbles::Bubbles,
+            EventCancelable::Cancelable,
+            None,
+            can_gc,
+        );
+        self.fire_clipboard_event(&event, action, can_gc);
```

Comment:
Can we add a comment that step 3 is handled by the VirtualMethods::handle_event of relevant elements that contain editable contexts?